client-sdk/ts-web/core/tests/hdkey.test.ts

```diff
@@ -0,0 +1,124 @@
+import * as bip39 from 'bip39';
+import * as nacl from 'tweetnacl'
+import {HdKey} from '../src/hdkey';
+import {concat} from '../src/misc';
+
+const toHex = (payload: Uint8Array | Buffer) => Buffer.from(payload).toString('hex');
+const hexToUint = (hex: string) => new Uint8Array(Buffer.from(hex, 'hex'));
+
+describe('HdKey', () => {
+    describe('#fromMnemonic', () => {
+        it('Should generate the root key from a mnemonic', async () => {
+            // Unsecure genre when ahead
+            const key = await HdKey.fromMnemonic('practice acoustic taxi')
+            expect(toHex(key.public())).toEqual('feb4ffe1b96066dd3c25ce1c468ee70e680c6f94d87dc3f158a2c310cbf37b74')
+        })
+
+        it('Should return null on unsecure mnemonics', async () => {
+            // Unsecure genre when ahead
+            const key = await HdKey.fromMnemonic('genre when ahead')
+            expect(key).toBeNull()
+        })
+    })
+
+    describe('#generateValidMnemonic', () => {
+        let spy: jest.SpyInstance;
+
+        it('Should retry until it finds a secure mnemonic', async () => {
+            spy = jest.spyOn(bip39, 'generateMnemonic')
+                // Generates an unsecure seed
+                .mockReturnValueOnce('genre when ahead')
+                // Generates a secure seed
+                .mockReturnValueOnce('practice acoustic taxi')
+
+            const key = await HdKey.generateValidMnemonic()
+            expect(spy).toHaveBeenCalledTimes(2)
+            expect(key).toBe('practice acoustic taxi')
+        })
+
+        afterEach(() => {
+            jest.restoreAllMocks()
+        })
+    })
+
+    describe('#fromSeed', () => {
+        it('Should throw with non-32 bytes seed', () => {
+            const tooShort = () => {
+                HdKey.fromSeed(new Uint8Array(31));
+            };
+            const tooLong = () => {
+                HdKey.fromSeed(new Uint8Array(33));
+            };
+
+            expect(tooShort).toThrowError(/must be 32 bytes/);
+            expect(tooLong).toThrowError(/must be 32 bytes/);
+        });
+
+        it('Should return node', () => {
+            const node = HdKey.fromSeed(new Uint8Array(32));
+
+            expect(toHex(node.A)).toBe(
+                '3b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29',
```

Comment:
Where are these test cases from? Would be great if we could use some standard test vectors in these tests.

Reply:
Saddly I could not find any test vectors for bip32-ed25519 derivation - so I made my own and checked them against a running oasis-node and verified the signed messages/pub keys using various libraries to cross-check my implementation

Reply:
Ok, then just provide a comment on how these were generated.